src/ray/common/ray_syncer/ray_syncer.h

```diff
@@ -0,0 +1,227 @@
+  /// \param node_id The id of current node.
+  RaySyncer(instrumented_io_context &io_context, const std::string &node_id);
+  ~RaySyncer();
+
+  /// Connect to a node.
+  /// TODO (iycheng): Introduce grpc channel pool and use node_id
+  /// for the connection.
+  ///
+  /// \param connection The connection to the remote node.
+  void Connect(std::unique_ptr<NodeSyncConnection> connection);
+
+  /// Connect to a node.
+  /// TODO (iycheng): Introduce grpc channel pool and use node_id
+  /// for the connection.
+  ///
+  /// \param connection The connection to the remote node.
+  void Connect(std::shared_ptr<grpc::Channel> channel);
+
+  void Disconnect(const std::string &node_id);
+
+  /// Register the components to the syncer module. Syncer will make sure eventually
+  /// it'll have a global view of the cluster.
+  ///
+  /// \param component_id The component to sync.
+  /// \param reporter The local component to be broadcasted.
+  /// \param receiver The snapshot of the component in the cluster.
+  /// \param upstream_only Only send the message to the upstream server for this
+  /// component. \param pull_from_reporter_interval_ms The frequence to pull a message
+  /// from reporter and push it to sending queue.
+  bool Register(RayComponentId component_id,
+                const ReporterInterface *reporter,
+                ReceiverInterface *receiver,
+                bool upstream_only = false,
+                int64_t pull_from_reporter_interval_ms = 100);
+
+  /// Function to broadcast the messages to other nodes.
+  /// A message will be sent to a node if that node doesn't have this message.
+  /// The message can be generated by local reporter or received by the other node.
+  ///
+  /// \param message The message to be broadcasted.
+  void BroadcastMessage(std::shared_ptr<const RaySyncMessage> message);
+
+  /// Get the current node id.
+  const std::string &GetNodeId() const { return node_id_; }
+
+  /// Get the io_context used by RaySyncer.
+  instrumented_io_context &GetIOContext() { return io_context_; }
+
+  /// Get the SyncConnection of a node.
+  ///
+  /// \param node_id The node id to lookup.
+  ///
+  /// \return nullptr if it doesn't exist, otherwise, the connection associated with the
+  /// node.
+  NodeSyncConnection *GetSyncConnection(const std::string &node_id) const {
+    auto iter = sync_connections_.find(node_id);
+    if (iter == sync_connections_.end()) {
+      return nullptr;
+    }
+    return iter->second.get();
+  }
+
+ private:
+  /// io_context for this thread
+  instrumented_io_context &io_context_;
+
+  /// The current node id.
+  const std::string node_id_;
+
+  /// Manage connections. Here the key is the NodeID in binary form.
+  absl::flat_hash_map<std::string, std::unique_ptr<NodeSyncConnection>> sync_connections_;
+
+  /// Upward connections. These are connections initialized not by this node
+  absl::flat_hash_set<NodeSyncConnection *> upward_connections_;
```

Comment:
Doesn't it make sense to unify the terms here? Use either client/server or upward/downward?

Reply:
This is for resource loads right now. I'm still thinking about a better way to support this. For resource loads, right now, we only push everything to one node. 

Btw, client and server are for connection. upward and downward is for direction.  I don't think it makes sense to unify the words here since they are for different things.